# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/nsRefreshDriver.cpp
# Commit: 1290c81b4fd7
# Full Hash: 1290c81b4fd788222a335a1bbe07747b28d22ae9
# Author: Alexander Surkov <surkov.alexander@gmail.com>
# Date: 2023-10-03 15:57:45
# Regressor Bug: 1807253
# File Overlap Count: 1
# Description:
#   Bug 1807253 - get rid of ResizeObserverNotificationHelper, r=emilio
#   
#   Integrate resize observer processing with RefreshDriver to make it
#   more extensible.
#   
# ==============================================================================

diff -r 5502331146fe -r 1290c81b4fd7 layout/base/nsRefreshDriver.cpp
--- a/layout/base/nsRefreshDriver.cpp	Tue Oct 03 09:40:06 2023 +0000
+++ b/layout/base/nsRefreshDriver.cpp	Tue Oct 03 09:48:46 2023 +0000
@@ -1334,7 +1334,8 @@
       mInNormalTick(false),
       mAttemptedExtraTickSinceLastVsync(false),
       mHasExceededAfterLoadTickPeriod(false),
-      mHasStartedTimerAtLeastOnce(false) {
+      mHasStartedTimerAtLeastOnce(false),
+      mResizeObservationCount(0) {
   MOZ_ASSERT(NS_IsMainThread());
   MOZ_ASSERT(mPresContext,
              "Need a pres context to tell us to call Disconnect() later "
@@ -1954,6 +1955,9 @@
   if (HasImageRequests() && !mThrottled) {
     reasons |= TickReasons::eHasImageRequests;
   }
+  if (mResizeObservationCount > 0) {
+    reasons |= TickReasons::eNeedsToNotifyResizeObservers;
+  }
   if (mNeedToUpdateIntersectionObservations) {
     reasons |= TickReasons::eNeedsToUpdateIntersectionObservations;
   }
@@ -1990,6 +1994,9 @@
   if (aReasons & TickReasons::eHasImageRequests) {
     aStr.AppendLiteral(" HasImageAnimations");
   }
+  if (mResizeObservationCount > 0) {
+    aStr.AppendLiteral(" NeedsToNotifyResizeObservers");
+  }
   if (aReasons & TickReasons::eNeedsToUpdateIntersectionObservations) {
     aStr.AppendLiteral(" NeedsToUpdateIntersectionObservations");
   }
@@ -2240,6 +2247,27 @@
   mNeedToUpdateContentRelevancy = false;
 }
 
+void nsRefreshDriver::NotifyResizeObservers() {
+  if (mResizeObservationCount == 0) {
+    return;
+  }
+
+  AutoTArray<RefPtr<Document>, 32> documents;
+
+  if (mPresContext->Document()->HasResizeObservers()) {
+    documents.AppendElement(mPresContext->Document());
+  }
+
+  mPresContext->Document()->CollectDescendantDocuments(
+      documents, [](const Document* document) -> bool {
+        return document->HasResizeObservers();
+      });
+
+  for (const RefPtr<Document>& doc : documents) {
+    MOZ_KnownLive(doc)->NotifyResizeObservers();
+  }
+}
+
 void nsRefreshDriver::DispatchAnimationEvents() {
   if (!mPresContext) {
     return;
@@ -2707,6 +2735,10 @@
     pm->UpdatePopupPositions(this);
   }
 
+  // Notify resize obsrevers if any, see
+  // https://html.spec.whatwg.org/#update-the-rendering step 14.
+  NotifyResizeObservers();
+
   // Update the relevancy of the content of any `content-visibility: auto`
   // elements. The specification says: "Specifically, such changes will
   // take effect between steps 13 and 14 of Update the Rendering step of